{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import shutil\n",
    "from Bio import SeqIO\n",
    "from itertools import combinations\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "work_path='/Users/jiachenghao/projects/Gal/eLife初稿/eLife_修回/Question4/Step2_Cal_HGT/'\n",
    "ARG_name='blaTEM-1B'\n",
    "inputname='China_Res_on_MGE'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "if os.path.exists(work_path+'Extract'):\n",
    "    Extract_fasta_file=work_path+'Extract/'\n",
    "else:\n",
    "    os.mkdir(work_path+'Extract/')\n",
    "    Extract_fasta_file=work_path+'Extract/'\n",
    "\n",
    "\n",
    "if os.path.exists(work_path+'Motifi_Extract_Fasta'):\n",
    "    Motifi=work_path+'Motifi_Extract_Fasta/'\n",
    "else:\n",
    "    os.mkdir(work_path+'Motifi_Extract_Fasta/')\n",
    "    Motifi=work_path+'Motifi_Extract_Fasta/'\n",
    "\n",
    "\n",
    "if os.path.exists(work_path+ARG_name+'_list'):\n",
    "    List_path=work_path+ARG_name+'_list/'\n",
    "else:\n",
    "    os.mkdir(work_path+ARG_name+'_list/')\n",
    "    List_path=work_path+ARG_name+'_list/'\n",
    "\n",
    "\n",
    "if os.path.exists(work_path+'ARG_file/'):\n",
    "    file_path=work_path+'ARG_file/'\n",
    "else:\n",
    "    os.mkdir(work_path+'ARG_file/')\n",
    "    file_path=work_path+'ARG_file/'\n",
    "\n",
    "\n",
    "if os.path.exists(work_path+'fasta_file/'):\n",
    "    fasta_path=work_path+'fasta_file/'\n",
    "else:\n",
    "    os.mkdir(work_path+'fasta_file/')\n",
    "    fasta_path=work_path+'fasta_file/'\n",
    "\n",
    "inputname=work_path+inputname+'/'   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def modify_contig_names(input_file, output_file, prefix):\n",
    "    with open(input_file, 'r') as f_in, open(output_file, 'w') as f_out:\n",
    "        for line in f_in:\n",
    "            if line.startswith('>'):\n",
    "                new_contig_name = f\">{prefix}\\n\"\n",
    "                f_out.write(new_contig_name)\n",
    "            else:\n",
    "                f_out.write(line)\n",
    "\n",
    "def merge_fasta_files(input_folder, output_file):\n",
    "    with open(output_file, 'w') as out_handle:\n",
    "        for filename in os.listdir(input_folder):\n",
    "            if filename.endswith('.fasta'):\n",
    "                file_path = os.path.join(input_folder, filename)\n",
    "                for record in SeqIO.parse(file_path, 'fasta'):\n",
    "                    SeqIO.write(record, out_handle, 'fasta')\n",
    "\n",
    "#定义函数，用于截取反向互补\n",
    "def rev(seq):\n",
    "    base_trans = {'A':'T', 'C':'G', 'T':'A', 'G':'C', 'a':'t', 'c':'g', 't':'a', 'g':'c'}\n",
    "    rev_seq = list(reversed(seq))\n",
    "    rev_seq_list = [base_trans[k] for k in rev_seq]\n",
    "    rev_seq = ''.join(rev_seq_list)\n",
    "    return(rev_seq)\n",
    "\n",
    "########################\n",
    "def extract_fasta(input, list, output):\n",
    "    seq_file = {}\n",
    "    with open(input, 'r') as input_fasta:\n",
    "        for line in input_fasta:\n",
    "            line = line.strip()\n",
    "            if line[0] == '>':\n",
    "                seq_id = line.split()[0]\n",
    "                seq_file[seq_id] = ''\n",
    "            else:\n",
    "                seq_file[seq_id] += line\n",
    "    input_fasta.close()\n",
    "\n",
    "    list_dict = {}\n",
    "    with open(list, 'r') as list_file:\n",
    "        for line in list_file:\n",
    "            if line.strip():\n",
    "                line = line.strip().split('\\t')\n",
    "                list_dict[line[0]] = [line[1], int(line[2]) - 1, int(line[3]), line[4]]\n",
    "    list_file.close()\n",
    "\n",
    "    output_fasta = open(output, 'w')\n",
    "    for key,value in list_dict.items():\n",
    "        print('>' + key, file = output_fasta)\n",
    "        seq = seq_file['>' + value[0]][value[1]:value[2]]\n",
    "        if value[3] == '+':\n",
    "            print(seq, file = output_fasta)\n",
    "        elif value[3] == '-':\n",
    "            seq = rev(seq)\n",
    "            print(seq, file = output_fasta)\n",
    "    output_fasta.close()\n",
    "    \n",
    "##########################\n",
    "    \n",
    "def remove_duplicates(strings):\n",
    "    seen_pairs = set()    #########\n",
    "    result = []\n",
    "    for string in strings:\n",
    "        parts = string.split('>')\n",
    "        pair = tuple(sorted(parts))\n",
    "        if pair not in seen_pairs:\n",
    "            seen_pairs.add(pair)\n",
    "            result.append(string)\n",
    "    return result\n",
    "\n",
    "def remove_duplicates2(strings):\n",
    "    result = []\n",
    "    for string in strings:\n",
    "        parts = string.split('>')\n",
    "        if len(parts) == 2 and parts[0] != parts[1]:\n",
    "            result.append(string)\n",
    "    return result\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Process_File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start process 227 files\n"
     ]
    }
   ],
   "source": [
    "inputname_list=os.listdir(inputname)\n",
    "inputname_list_fna=[x for x in inputname_list if x.endswith('.fasta')]\n",
    "inputname_list_txt=[x for x in inputname_list if x.endswith('.csv')]\n",
    "if len(inputname_list_fna) == len(inputname_list_txt):\n",
    "    print('Start process '+ str(len(inputname_list_txt)) + ' files' )\n",
    "for i in inputname_list_txt:\n",
    "    shutil.copy(inputname+i,file_path)\n",
    "for i in inputname_list_fna:\n",
    "    shutil.copy(inputname+i,fasta_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_txt=work_path+'All_txt.txt'\n",
    "with open(merged_txt, 'w') as outfile:\n",
    "    for filename in os.listdir(file_path):\n",
    "        if filename.endswith('.csv'):\n",
    "            temp_file_path = os.path.join(file_path, filename)\n",
    "            with open(temp_file_path, 'r') as infile:\n",
    "                outfile.write(infile.read())\n",
    "file=pd.read_csv(merged_txt)\n",
    "file=file[file['PRODUCT']==ARG_name]\n",
    "\n",
    "\n",
    "file.to_csv(work_path+'Target_ARG_summary.csv',index=False)\n",
    "\n",
    "file=pd.read_csv('Target_ARG_summary.csv')\n",
    "file_length=file.shape[0]\n",
    "os.remove(merged_txt)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Start Extract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "counter=0\n",
    "while counter < file_length:\n",
    "    txt_out=[]\n",
    "    temp_file=file[file.index==counter]\n",
    "    fasta_list_filename=temp_file['#FILE'].to_list()[0]\n",
    "    \n",
    "    ARG_name=temp_file['PRODUCT'].to_list()[0]\n",
    "\n",
    "    SEQUENCE=temp_file['SEQUENCE'].to_list()[0]\n",
    "\n",
    "    START=temp_file['START'].to_list()[0]\n",
    "    START=str(START)\n",
    "\n",
    "    END=temp_file['END'].to_list()[0]\n",
    "    END=str(END)\n",
    "\n",
    "    STRAND=temp_file['STRAND'].to_list()[0]\n",
    "\n",
    "    txt_out.append(SEQUENCE)\n",
    "    txt_out.append(START)\n",
    "    txt_out.append(END)\n",
    "    txt_out.append(STRAND)\n",
    "\n",
    "    if len(txt_out)==4:\n",
    "        txt_out=pd.DataFrame(txt_out)\n",
    "        txt_out=txt_out.T\n",
    "        txt_out.to_csv(List_path+fasta_list_filename+'_'+str(counter),sep='\\t',header=False)\n",
    "    counter+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SAL03426.fasta\n",
      "Extracting SAL03426.fasta_40\n",
      "SAL02648.fasta\n",
      "Extracting SAL02648.fasta_154\n",
      "SAL04098.fasta\n",
      "Extracting SAL04098.fasta_114\n",
      "SAL02681.fasta\n",
      "Extracting SAL02681.fasta_153\n",
      "SAL02652.fasta\n",
      "Extracting SAL02652.fasta_3\n",
      "SAL_PA7723AA.fasta\n",
      "Extracting SAL_PA7723AA.fasta_123\n",
      "SAL03413.fasta\n",
      "Extracting SAL03413.fasta_46\n",
      "SAL02650.fasta\n",
      "Extracting SAL02650.fasta_95\n",
      "SAL03419.fasta\n",
      "Extracting SAL03419.fasta_47\n",
      "SAL03307.fasta\n",
      "Extracting SAL03307.fasta_137\n",
      "SAL00718.fasta\n",
      "Extracting SAL00718.fasta_124\n",
      "SAL02651.fasta\n",
      "Extracting SAL02651.fasta_139\n",
      "SAL03593.fasta\n",
      "Extracting SAL03593.fasta_113\n",
      "SAL03416.fasta\n",
      "Extracting SAL03416.fasta_152\n",
      "SAL03406.fasta\n",
      "Extracting SAL03406.fasta_87\n",
      "SAL02670.fasta\n",
      "Extracting SAL02670.fasta_136\n",
      "SAL03448.fasta\n",
      "Extracting SAL03448.fasta_35\n",
      "SAL03427.fasta\n",
      "Extracting SAL03427.fasta_79\n",
      "SAL02677.fasta\n",
      "Extracting SAL02677.fasta_115\n",
      "SAL00708.fasta\n",
      "Extracting SAL00708.fasta_9\n",
      "SAL03306.fasta\n",
      "Extracting SAL03306.fasta_94\n",
      "SAL_PA7729AA.fasta\n",
      "Extracting SAL_PA7729AA.fasta_155\n",
      "SAL03300.fasta\n",
      "Extracting SAL03300.fasta_112\n",
      "SAL03288.fasta\n",
      "Extracting SAL03288.fasta_138\n",
      "SAL_PA7701AA.fasta\n",
      "Extracting SAL_PA7701AA.fasta_99\n",
      "SAL03282.fasta\n",
      "Extracting SAL03282.fasta_125\n",
      "SAL02673.fasta\n",
      "Extracting SAL02673.fasta_6\n",
      "SAL_PA7686AA.fasta\n",
      "Extracting SAL_PA7686AA.fasta_11\n",
      "TH257.fasta\n",
      "Extracting TH257.fasta_29\n",
      "SAL03296.fasta\n",
      "Extracting SAL03296.fasta_122\n",
      "SAL03421.fasta\n",
      "Extracting SAL03421.fasta_62\n",
      "SAL03298.fasta\n",
      "Extracting SAL03298.fasta_76\n",
      "SAL03330.fasta\n",
      "Extracting SAL03330.fasta_53\n",
      "SAL03418.fasta\n",
      "Extracting SAL03418.fasta_1\n",
      "SAL02647.fasta\n",
      "Extracting SAL02647.fasta_50\n",
      "SAL03315.fasta\n",
      "Extracting SAL03315.fasta_163\n",
      "SAL03409.fasta\n",
      "Extracting SAL03409.fasta_164\n",
      "SAL04097.fasta\n",
      "Extracting SAL04097.fasta_36\n",
      "SAL03328.fasta\n",
      "Extracting SAL03328.fasta_20\n",
      "SAL03439.fasta\n",
      "Extracting SAL03439.fasta_5\n",
      "SAL03304.fasta\n",
      "Extracting SAL03304.fasta_2\n",
      "SAL03286.fasta\n",
      "Extracting SAL03286.fasta_18\n",
      "SAL02660.fasta\n",
      "Extracting SAL02660.fasta_71\n",
      "SAL_PA7695AA.fasta\n",
      "Extracting SAL_PA7695AA.fasta_51\n",
      "SAL03292.fasta\n",
      "Extracting SAL03292.fasta_13\n",
      "SAL_PA7730AA.fasta\n",
      "Extracting SAL_PA7730AA.fasta_165\n",
      "SAL02674.fasta\n",
      "Extracting SAL02674.fasta_73\n",
      "SAL04028.fasta\n",
      "Extracting SAL04028.fasta_74\n",
      "TH254.fasta\n",
      "Extracting TH254.fasta_162\n",
      "SAL03436.fasta\n",
      "Extracting SAL03436.fasta_98\n",
      "SAL04093.fasta\n",
      "Extracting SAL04093.fasta_100\n",
      "SAL04031.fasta\n",
      "Extracting SAL04031.fasta_78\n",
      "SAL_UA3243AA.fasta\n",
      "Extracting SAL_UA3243AA.fasta_60\n",
      "SAL_PA7725AA.fasta\n",
      "Extracting SAL_PA7725AA.fasta_177\n",
      "SAL03417.fasta\n",
      "Extracting SAL03417.fasta_107\n",
      "SAL00735.fasta\n",
      "Extracting SAL00735.fasta_171\n",
      "SAL00725.fasta\n",
      "Extracting SAL00725.fasta_52\n",
      "SAL04094.fasta\n",
      "Extracting SAL04094.fasta_179\n",
      "SAL02668.fasta\n",
      "Extracting SAL02668.fasta_101\n",
      "SAL_PA7976AA.fasta\n",
      "Extracting SAL_PA7976AA.fasta_12\n",
      "SAL00738.fasta\n",
      "Extracting SAL00738.fasta_178\n",
      "SAL00737.fasta\n",
      "Extracting SAL00737.fasta_72\n",
      "SAL04099.fasta\n",
      "Extracting SAL04099.fasta_170\n",
      "SAL03425.fasta\n",
      "Extracting SAL03425.fasta_176\n",
      "SAL03388.fasta\n",
      "Extracting SAL03388.fasta_80\n",
      "SAL02658.fasta\n",
      "Extracting SAL02658.fasta_85\n",
      "SAL03403.fasta\n",
      "Extracting SAL03403.fasta_106\n",
      "SAL02654.fasta\n",
      "Extracting SAL02654.fasta_31\n",
      "SAL02653.fasta\n",
      "Extracting SAL02653.fasta_48\n",
      "SAL03437.fasta\n",
      "Extracting SAL03437.fasta_148\n",
      "TH255.fasta\n",
      "Extracting TH255.fasta_108\n",
      "SAL04035.fasta\n",
      "Extracting SAL04035.fasta_146\n",
      "C15_FKDN220483769-1A_H25LVDSX5_L3.clean.fq.fasta\n",
      "Extracting C15_FKDN220483769-1A_H25LVDSX5_L3.clean.fq.fasta_67\n",
      "SAL_PA7719AA.fasta\n",
      "Extracting SAL_PA7719AA.fasta_10\n",
      "SAL04020.fasta\n",
      "Extracting SAL04020.fasta_97\n",
      "SAL04026.fasta\n",
      "Extracting SAL04026.fasta_131\n",
      "SAL02698.fasta\n",
      "Extracting SAL02698.fasta_140\n",
      "SAL02702.fasta\n",
      "Extracting SAL02702.fasta_91\n",
      "SAL00744.fasta\n",
      "Extracting SAL00744.fasta_147\n",
      "SAL02699.fasta\n",
      "Extracting SAL02699.fasta_96\n",
      "SAL03434.fasta\n",
      "Extracting SAL03434.fasta_19\n",
      "SAL04092.fasta\n",
      "Extracting SAL04092.fasta_149\n",
      "XXB1092.fasta\n",
      "Extracting XXB1092.fasta_22\n",
      "SAL03314.fasta\n",
      "Extracting SAL03314.fasta_109\n",
      "SAL03430.fasta\n",
      "Extracting SAL03430.fasta_130\n",
      "SAL04019.fasta\n",
      "Extracting SAL04019.fasta_141\n",
      "E404.clean.fq.fasta\n",
      "Extracting E404.clean.fq.fasta_7\n",
      "SAL03420.fasta\n",
      "Extracting SAL03420.fasta_23\n",
      "SAL02700.fasta\n",
      "Extracting SAL02700.fasta_0\n",
      "SAL03408.fasta\n",
      "Extracting SAL03408.fasta_110\n",
      "SAL03433.fasta\n",
      "Extracting SAL03433.fasta_84\n",
      "SAL04018.fasta\n",
      "Extracting SAL04018.fasta_93\n",
      "SAL03291.fasta\n",
      "Extracting SAL03291.fasta_157\n",
      "SAL03386.fasta\n",
      "Extracting SAL03386.fasta_121\n",
      "SAL00739.fasta\n",
      "Extracting SAL00739.fasta_127\n",
      "SAL03414.fasta\n",
      "Extracting SAL03414.fasta_59\n",
      "SAL02665.fasta\n",
      "Extracting SAL02665.fasta_90\n",
      "SAL02663.fasta\n",
      "Extracting SAL02663.fasta_119\n",
      "SAL02664.fasta\n",
      "Extracting SAL02664.fasta_133\n",
      "TH260.fasta\n",
      "Extracting TH260.fasta_117\n",
      "SAL_PA7698AA.fasta\n",
      "Extracting SAL_PA7698AA.fasta_150\n",
      "SAL_PA7759AA.fasta\n",
      "Extracting SAL_PA7759AA.fasta_159\n",
      "SAL00743.fasta\n",
      "Extracting SAL00743.fasta_129\n",
      "SAL03451.fasta\n",
      "Extracting SAL03451.fasta_26\n",
      "SAL03326.fasta\n",
      "Extracting SAL03326.fasta_134\n",
      "SAL02656.fasta\n",
      "Extracting SAL02656.fasta_111\n",
      "SAL03411.fasta\n",
      "Extracting SAL03411.fasta_126\n",
      "SAL03312.fasta\n",
      "Extracting SAL03312.fasta_92\n",
      "SAL04095.fasta\n",
      "Extracting SAL04095.fasta_128\n",
      "ERS2630811.fasta\n",
      "Extracting ERS2630811.fasta_44\n",
      "TH256.fasta\n",
      "Extracting TH256.fasta_77\n",
      "SAL03440.fasta\n",
      "Extracting SAL03440.fasta_135\n",
      "SAL_GB9917AA.fasta\n",
      "Extracting SAL_GB9917AA.fasta_158\n",
      "C13_FKDN220485113-1A_H25L7DSX5_L2.clean.fq.fasta\n",
      "Extracting C13_FKDN220485113-1A_H25L7DSX5_L2.clean.fq.fasta_151\n",
      "SAL03435.fasta\n",
      "Extracting SAL03435.fasta_63\n",
      "SAL03402.fasta\n",
      "Extracting SAL03402.fasta_156\n",
      "SAL03429.fasta\n",
      "Extracting SAL03429.fasta_118\n",
      "SAL00734.fasta\n",
      "Extracting SAL00734.fasta_116\n",
      "SAL03453.fasta\n",
      "Extracting SAL03453.fasta_120\n",
      "SAL03407.fasta\n",
      "Extracting SAL03407.fasta_41\n",
      "SAL03385.fasta\n",
      "Extracting SAL03385.fasta_70\n",
      "SAL03309.fasta\n",
      "Extracting SAL03309.fasta_16\n",
      "SAL00717.fasta\n",
      "Extracting SAL00717.fasta_32\n",
      "SAL02678.fasta\n",
      "Extracting SAL02678.fasta_38\n",
      "SAL02657.fasta\n",
      "Extracting SAL02657.fasta_160\n",
      "SAL02672.fasta\n",
      "Extracting SAL02672.fasta_55\n",
      "SAL03452.fasta\n",
      "Extracting SAL03452.fasta_169\n",
      "SAL03303.fasta\n",
      "Extracting SAL03303.fasta_75\n",
      "SAL02662.fasta\n",
      "Extracting SAL02662.fasta_167\n",
      "SAL04029.fasta\n",
      "Extracting SAL04029.fasta_25\n",
      "C49_FKDN220483803-1A_H25LVDSX5_L1.clean.fq.fasta\n",
      "Extracting C49_FKDN220483803-1A_H25LVDSX5_L1.clean.fq.fasta_33\n",
      "SRR26662937.fastq.fasta\n",
      "Extracting SRR26662937.fastq.fasta_37\n",
      "SAL03301.fasta\n",
      "Extracting SAL03301.fasta_161\n",
      "SAL_PA7700AA.fasta\n",
      "Extracting SAL_PA7700AA.fasta_30\n",
      "SAL04017.fasta\n",
      "Extracting SAL04017.fasta_15\n",
      "SAL03397.fasta\n",
      "Extracting SAL03397.fasta_57\n",
      "SAL02661.fasta\n",
      "Extracting SAL02661.fasta_28\n",
      "SAL03325.fasta\n",
      "Extracting SAL03325.fasta_4\n",
      "SAL02675.fasta\n",
      "Extracting SAL02675.fasta_24\n",
      "SAL04101.fasta\n",
      "Extracting SAL04101.fasta_27\n",
      "SAL04091.fasta\n",
      "Extracting SAL04091.fasta_21\n",
      "SAL02701.fasta\n",
      "Extracting SAL02701.fasta_49\n",
      "SAL03428.fasta\n",
      "Extracting SAL03428.fasta_166\n",
      "SAL_PA7696AA.fasta\n",
      "Extracting SAL_PA7696AA.fasta_43\n",
      "SAL04102.fasta\n",
      "Extracting SAL04102.fasta_168\n",
      "XXB1725.fasta\n",
      "Extracting XXB1725.fasta_56\n",
      "SAL03431.fasta\n",
      "Extracting SAL03431.fasta_180\n",
      "SAL04025.fasta\n",
      "Extracting SAL04025.fasta_83\n",
      "SAL02649.fasta\n",
      "Extracting SAL02649.fasta_104\n",
      "SAL03293.fasta\n",
      "Extracting SAL03293.fasta_61\n",
      "SAL00746.fasta\n",
      "Extracting SAL00746.fasta_65\n",
      "SAL02676.fasta\n",
      "Extracting SAL02676.fasta_172\n",
      "SAL02659.fasta\n",
      "Extracting SAL02659.fasta_45\n",
      "SAL03297.fasta\n",
      "Extracting SAL03297.fasta_174\n",
      "SAL03302.fasta\n",
      "Extracting SAL03302.fasta_34\n",
      "SAL04090.fasta\n",
      "Extracting SAL04090.fasta_66\n",
      "SAL03294.fasta\n",
      "Extracting SAL03294.fasta_42\n",
      "SAL03422.fasta\n",
      "Extracting SAL03422.fasta_103\n",
      "SAL03308.fasta\n",
      "Extracting SAL03308.fasta_58\n",
      "SAL03329.fasta\n",
      "Extracting SAL03329.fasta_64\n",
      "TH261.fasta\n",
      "Extracting TH261.fasta_173\n",
      "SAL02667.fasta\n",
      "Extracting SAL02667.fasta_8\n",
      "SAL03290.fasta\n",
      "Extracting SAL03290.fasta_105\n",
      "SAL04096.fasta\n",
      "Extracting SAL04096.fasta_81\n",
      "SAL04034.fasta\n",
      "Extracting SAL04034.fasta_102\n",
      "TH259.fasta\n",
      "Extracting TH259.fasta_175\n",
      "SAL03295.fasta\n",
      "Extracting SAL03295.fasta_86\n",
      "SAL04100.fasta\n",
      "Extracting SAL04100.fasta_68\n",
      "SAL04032.fasta\n",
      "Extracting SAL04032.fasta_132\n",
      "SAL03401.fasta\n",
      "Extracting SAL03401.fasta_14\n",
      "SAL03449.fasta\n",
      "Extracting SAL03449.fasta_82\n",
      "SAL03450.fasta\n",
      "Extracting SAL03450.fasta_69\n",
      "SAL03423.fasta\n",
      "Extracting SAL03423.fasta_145\n",
      "SAL_PA7745AA.fasta\n",
      "Extracting SAL_PA7745AA.fasta_142\n",
      "SAL02671.fasta\n",
      "Extracting SAL02671.fasta_89\n",
      "SAL03432.fasta\n",
      "Extracting SAL03432.fasta_39\n",
      "SAL03327.fasta\n",
      "Extracting SAL03327.fasta_88\n",
      "SAL03415.fasta\n",
      "Extracting SAL03415.fasta_17\n",
      "SAL03438.fasta\n",
      "Extracting SAL03438.fasta_54\n",
      "SAL02669.fasta\n",
      "Extracting SAL02669.fasta_143\n",
      "SAL03615.fasta\n",
      "Extracting SAL03615.fasta_144\n"
     ]
    }
   ],
   "source": [
    "List_path_file=os.listdir(List_path)\n",
    "for i in List_path_file:\n",
    "    last_underscore_index=i.rfind('_')\n",
    "    real_fasta_name=i[:last_underscore_index]\n",
    "    print(real_fasta_name)\n",
    "\n",
    "    print('Extracting '+i)\n",
    "    input_fasta=fasta_path+real_fasta_name\n",
    "    list_file=List_path+i\n",
    "    output=Extract_fasta_file+i\n",
    "    extract_fasta(input_fasta,list_file,output)\n",
    "\n",
    "fastas=os.listdir(Extract_fasta_file)\n",
    "for i in fastas:\n",
    "    contig_name=i.replace('.fasta','')\n",
    "    modify_contig_names(Extract_fasta_file+i, Motifi+i,contig_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "########多一步 Mutil Fasta Files合并########\n",
    "Motified_fastas=os.listdir(Motifi)\n",
    "temp_Motified=[]\n",
    "for i in Motified_fastas:\n",
    "    last_underscore_index=i.rfind('_')\n",
    "    temp_temp_Motified=i[:last_underscore_index]\n",
    "    temp_Motified.append(temp_temp_Motified)\n",
    "\n",
    "set_temp_Motified=list(set(temp_Motified))\n",
    "for i in set_temp_Motified:\n",
    "\n",
    "    temp_count=temp_Motified.count(i)\n",
    "\n",
    "    if temp_count > 1:\n",
    "        temp_Motified_file_list=[x for x in Motified_fastas if x.startswith(i)]\n",
    "\n",
    "        with open(Motifi+i, 'w') as outfile:\n",
    "            for ii in temp_Motified_file_list:\n",
    "                with open(Motifi+ii, 'r') as infile:\n",
    "                    for line in infile:\n",
    "                        outfile.write(line)\n",
    "    else:\n",
    "        temp_Motified_file_list=[x for x in Motified_fastas if x.startswith(i)][0]\n",
    "        os.rename(Motifi+temp_Motified_file_list,Motifi+i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "181 Strain carried blaTEM-1B\n"
     ]
    }
   ],
   "source": [
    "Motified_fastas=os.listdir(Motifi)\n",
    "\n",
    "for i in Motified_fastas:\n",
    "    ii = os.path.join(Motifi, i)\n",
    "    if os.path.isfile(ii) and not i.endswith(\".fasta\"):\n",
    "        os.remove(ii)\n",
    "######传递参数######\n",
    "Key_Value=os.listdir(Motifi) \n",
    "Key_Value=len(Key_Value)\n",
    "print(Key_Value, 'Strain carried', ARG_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_file=ARG_name+' merged.fasta'\n",
    "merge_fasta_files(Motifi, output_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "构建Blastn文件结构"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/jiachenghao/projects/Gal/eLife初稿/eLife_修回/Question4/Step2_Cal_HGT/blastn/merged_DB/merged.fasta'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if os.path.exists(work_path+'blastn'):\n",
    "    blastn_path=work_path+'blastn'\n",
    "else:\n",
    "    os.mkdir(work_path+'blastn')\n",
    "    blastn_path=work_path+'blastn'\n",
    "\n",
    "if os.path.exists(work_path+output_file):\n",
    "    shutil.move(work_path+output_file,blastn_path)\n",
    "else:\n",
    "    print(output_file, \" Missing\")\n",
    "\n",
    "if os.path.exists(Motifi):\n",
    "    shutil.move(Motifi,blastn_path)\n",
    "else:\n",
    "    print(\"Motifi folder Missing\")\n",
    "\n",
    "if os.path.exists(work_path+'blastn.sh'):\n",
    "    shutil.copy(work_path+'blastn.sh',blastn_path)\n",
    "else:\n",
    "    print(\"Blatn Script Missing\")\n",
    "\n",
    "os.mkdir(blastn_path+'/OUT')\n",
    "os.mkdir(blastn_path+'/merged_DB')\n",
    "os.rename(blastn_path+'/'+output_file, blastn_path+'/'+'merged.fasta')\n",
    "shutil.move(blastn_path+'/merged.fasta',blastn_path+'/merged_DB')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "构建blastn索引"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Blastn DB Run Success\n",
      "\n",
      "\n",
      "Building a new DB, current time: 10/17/2024 10:44:32\n",
      "New DB name:   /Users/jiachenghao/projects/Gal/eLife初稿/eLife_修回/Question4/Step2_Cal_HGT/blastn/merged_DB/merged\n",
      "New DB title:  merged.fasta\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 181 sequences in 0.00478601 seconds.\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "working_directory=blastn_path+'/merged_DB'\n",
    "command = [\"makeblastdb\", \"-in\", \"merged.fasta\", \"-dbtype\", \"nucl\", \"-out\", \"merged\", \"-parse_seqids\"]\n",
    "result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, cwd=working_directory)\n",
    "if result.returncode == 0:\n",
    "    print(\"Blastn DB Run Success\")\n",
    "    print(result.stdout.decode())\n",
    "else:\n",
    "    print(\"Blastn DB Run Failed\")\n",
    "    print(result.stderr.decode())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "运行Blastn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Blastn Run Complete\n",
      "\n"
     ]
    }
   ],
   "source": [
    "working_directory=blastn_path\n",
    "command = [\"sh\", \"blastn.sh\"]\n",
    "result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, cwd=working_directory)\n",
    "\n",
    "# 输出命令执行结果\n",
    "if result.returncode == 0:\n",
    "    print(\"Blastn Run Complete\")\n",
    "    print(result.stdout.decode())\n",
    "else:\n",
    "    print(\"Blastn Run Failed\")\n",
    "    print(result.stderr.decode())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "计算HGT Frequent Percentage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File Combined\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/Users/jiachenghao/projects/Gal/eLife初稿/eLife_修回/Question4/Step2_Cal_HGT/Final_blastn_merged.txt'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "working_directory=blastn_path+'/OUT'\n",
    "command = [\"cat\", \"*.txt\", \">\", \"Final_blastn_merged.txt\"]\n",
    "result = subprocess.run(\" \".join(command), shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, cwd=working_directory)\n",
    "\n",
    "if result.returncode == 0:\n",
    "    print(\"File Combined\")\n",
    "    print(result.stdout.decode())\n",
    "else:\n",
    "    print(\"File Combined Failed\")\n",
    "    print(result.stderr.decode())\n",
    "\n",
    "shutil.move(blastn_path+'/OUT/Final_blastn_merged.txt',work_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "######下面重写"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "181 Starin carry a total of 181 blaTEM-1B\n"
     ]
    }
   ],
   "source": [
    "file=pd.read_csv('Final_blastn_merged.txt',sep='\\t',header=None)\n",
    "column=['name1','name2','name3','name4','name5','name6','name7','name8','name9','name10',\n",
    "        'name11','name12','name13','name14']\n",
    "file.columns=column\n",
    "file_length2=file.shape[0]\n",
    "x_col=['>']*file_length2\n",
    "file['x']=x_col\n",
    "file['y']=file.apply(lambda row: ''.join(row[['name1', 'x', 'name2']]), axis=1)\n",
    "\n",
    "file_name1=list(set(file['name1'].to_list()))\n",
    "print(len(file_name1), 'Starin carry a total of',len(os.listdir(Extract_fasta_file)),ARG_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The 100% Coverage is 861\n"
     ]
    }
   ],
   "source": [
    "All_Y=[]\n",
    "All_verify=[]\n",
    "Check_MAX=[]\n",
    "\n",
    "for i in file_name1:\n",
    "    temp_file=file[file['name1']==i]\n",
    "\n",
    "    new_index=i+'>'+i\n",
    "    temp_file2=temp_file[temp_file['y']==new_index]\n",
    "    temp_file2_name4=temp_file2['name4'].to_list()\n",
    "    max_value = max(temp_file2_name4)   #### Coverage == 100%\n",
    "    Check_MAX.append(max_value)\n",
    "\n",
    "    temp_file3=temp_file[temp_file['name4']==max_value]\n",
    "    temp_file3=temp_file3[temp_file3['name3'] ==  100]  #### Identity == 100%\n",
    "    \n",
    "    ready_to_depulicate=temp_file3['y'].to_list()\n",
    "    verfy=temp_file3['name4'].to_list()\n",
    "    for ii in ready_to_depulicate:\n",
    "        All_Y.append(ii)\n",
    "    for iii in verfy:\n",
    "        All_verify.append(iii)\n",
    "\n",
    "\n",
    "print('The 100% Coverage is', list(set(Check_MAX))[0])\n",
    "if len(list(set(Check_MAX))) > 1:\n",
    "    print('Warning: the 100% Identity is not only.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_1=remove_duplicates(All_Y)\n",
    "filtered_2=remove_duplicates2(filtered_1)\n",
    "temp_out=pd.DataFrame(filtered_2)\n",
    "temp_out.to_csv('HGT_Detailed.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "############All_Y#############\n",
    "All_Y_Motified=[]\n",
    "# test=['GMBC10.085_118_34-GMBC10.027_724_19']\n",
    "for i in All_Y:\n",
    "    i=i.split('>')\n",
    "    i_motifi_1=i[0]\n",
    "    last_underscore_index=i_motifi_1.rfind('_')\n",
    "    i_motifi_1=i_motifi_1[:last_underscore_index]\n",
    "\n",
    "    i_motifi_2=i[-1]\n",
    "    last_underscore_index=i_motifi_2.rfind('_')\n",
    "    i_motifi_2=i_motifi_2[:last_underscore_index]\n",
    "\n",
    "    Motified_i=i_motifi_1+'>'+i_motifi_2\n",
    "    All_Y_Motified.append(Motified_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_1=remove_duplicates(All_Y_Motified)\n",
    "filtered_2=remove_duplicates2(filtered_1)\n",
    "temp_out=pd.DataFrame(filtered_2)\n",
    "temp_out.to_csv('HGT_Strain_Non_Detailed.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# file_length=len(os.listdir(fasta_path))  ###这个地方可能错误了\n",
    "file_length=Key_Value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16290\n",
      "16290.0\n",
      "HGT Frequency for  blaTEM-1B is: 1.0\n"
     ]
    }
   ],
   "source": [
    "# final_all_all=file_length*file_length    #######Bug2\n",
    "final_all_count=len(filtered_2)\n",
    "print(final_all_count)\n",
    "final_all_all=(1+(file_length-1))*((file_length-1)/2)\n",
    "print(final_all_all) \n",
    "HGT_Frequency=final_all_count/final_all_all\n",
    "HGT_Frequency=round(HGT_Frequency,2)\n",
    "print('HGT Frequency for ', ARG_name, 'is:', HGT_Frequency)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Deleting Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "shutil.rmtree(work_path+'blastn')\n",
    "shutil.rmtree(work_path+'Extract')\n",
    "shutil.rmtree(work_path+'fasta_file')\n",
    "shutil.rmtree(work_path+'ARG_file')\n",
    "shutil.rmtree(work_path+ARG_name+'_list/')\n",
    "os.remove(work_path+'Final_blastn_merged.txt')\n",
    "os.remove(work_path+'Target_ARG_summary.csv')\n",
    "os.remove(work_path+'HGT_Strain_Non_Detailed.csv')\n",
    "os.remove(work_path+'HGT_Detailed.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jia_node05",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
